src/x509.lisp

```diff
@@ -0,0 +1,358 @@
+  (let* ((data (asn1-string-data asn1-string))
+         (length (asn1-string-length asn1-string))
+         (vector (cffi:make-shareable-byte-vector length)))
+    (copy-bytes-to-lisp-vector data vector length)
+    vector))
+
+(declaim (inline asn1-iastring-char-p))
+(defun asn1-iastring-char-p (byte)
+  (declare (type (unsigned-byte 8) byte)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (< byte #x80))
+
+(defun asn1-iastring-p (bytes)
+  (declare (type (simple-array (unsigned-byte 8)) bytes)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (every #'asn1-iastring-char-p bytes))
+
+(defmethod decode-asn1-string (asn1-string (type (eql #.+v-asn1-iastring+)))
+  (let* ((data (asn1-string-data asn1-string))
+         (length (asn1-string-length asn1-string))
+         (strlen (cffi::foreign-string-length data :encoding :ascii)))
+    (if (= strlen length)
+        (let ((bytes (asn1-string-bytes-vector asn1-string)))
+          (if (asn1-iastring-p bytes)
+              (cffi:foreign-string-to-lisp data :encoding :ascii)
+              (error 'invalid-asn1-string :type '+v-asn1-iastring+)))
+        (error 'invalid-asn1-string :type '+v-asn1-iastring+))))
+
+(declaim (inline asn1-printable-char-p))
+(defun asn1-printable-char-p (byte)
+  (declare (type (unsigned-byte 8) byte)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (cond
+    ;; a-z
+    ((and (>= byte #.(char-code #\a))
+          (<= byte #.(char-code #\z)))
+     t)
+    ;; '-/
+    ((and (>= byte #.(char-code #\'))
+          (<= byte #.(char-code #\/)))
+     t)
+    ;; 0-9
+    ((and (>= byte #.(char-code #\0))
+          (<= byte #.(char-code #\9)))
+     t)
+    ;; A-Z
+    ((and (>= byte #.(char-code #\A))
+          (<= byte #.(char-code #\Z)))
+     t)
+    ;; other
+    ((= byte #.(char-code #\ )) t)
+    ((= byte #.(char-code #\:)) t)
+    ((= byte #.(char-code #\=)) t)
+    ((= byte #.(char-code #\?)) t)))
+
+(defun asn1-printable-string-p (bytes)
+  (declare (type (simple-array (unsigned-byte 8)) bytes)
+           (optimize (speed 3)
+                     (debug 0)
+                     (safety 0)))
+  (every #'asn1-printable-char-p bytes))
+
+(defmethod decode-asn1-string (asn1-string (type (eql #.+v-asn1-printablestring+)))
+  (let* ((bytes (asn1-string-bytes-vector asn1-string)))
+    (if (asn1-printable-string-p bytes)
+        (babel:octets-to-string bytes :encoding :ascii)
+        (error 'invalid-asn1-string :type '+v-asn1-printablestring+))))
+
+(defmethod decode-asn1-string (asn1-string (type (eql #.+v-asn1-utf8string+)))
+  (let* ((data (asn1-string-data asn1-string))
+         (length (asn1-string-length asn1-string))
+         (strlen (cffi::foreign-string-length data :encoding :ascii)))
+    (if (= strlen length)
+        (cffi:foreign-string-to-lisp data)
```

Comment:
Also, taking into account that "In general it cannot be assumed that the data returned by ASN1_STRING_data() is null terminated " [1], IMHO we need to pass :count : `(cffi:foreign-string-to-lisp data :encoding :utf-8 :count length)`

[1] - https://www.openssl.org/docs/manmaster/crypto/ASN1_STRING_dup.html